WGSL shader conversion for the examples

kvark opened the pull request:
Blocked on:
  - ~~https://github.com/gfx-rs/naga/issues/335~~
  - ~~https://github.com/gfx-rs/naga/issues/336~~
  - ~~https://github.com/gfx-rs/wgpu/issues/1095~~
  - ~~https://github.com/gfx-rs/naga/pull/360~~
  - https://github.com/gfx-rs/gfx/pull/3622
  - ~~GL backend doing the program link check earlier~~

Converted:
- [x] boids
- [x] cube
- [x] mipmap
- [x] msaa-line
- [x] shadow
- [x] skybox
- [ ] texture-array
- [ ] water

baptistemanson commented:
I tested all examples on develop and on this branch on Windows 10 using a RTX 2080 SUPER.
I have 3 regressions:

Cube seems to have some coords issues:
![image](https://user-images.githubusercontent.com/5444992/105647210-d9414b00-5e58-11eb-93dd-22665ad4648c.png)


Hello triangle
thread 'main' panicked at 'Failed to find an appropriate adapter', examples\hello-triangle\main.rs:19:10

MipMap doesn't display
![image](https://user-images.githubusercontent.com/5444992/105647227-fb3acd80-5e58-11eb-9af9-f1757c095c20.png)

They all work properly on origin/develop.

I'll dig more into those later!

baptistemanson commented:
Thanks for the force push, it solved the hello-triangle issue.
When removing the projection matrix from the cube example, everything works as expected. I suspect changing from GLSL to WGSL has maybe change the convention around matrices (row first <> column first or something similar?)
![image](https://user-images.githubusercontent.com/5444992/105647975-e90f5e00-5e5d-11eb-8553-526aabd92ec2.png)

grovesNL commented on examples/cube/main.rs:
```diff
@@ -274,19 +271,27 @@ impl framework::Example for Example {
 
-        let vs_module = device.create_shader_module(&wgpu::include_spirv!("shader.vert.spv"));
-        let fs_module = device.create_shader_module(&wgpu::include_spirv!("shader.frag.spv"));
+        let mut flags = wgpu::ShaderFlags::VALIDATION;
```
nit: we could temporarily have a helper function for this in framework to remove backend matching from the example code

kvark replied:
I considered this, but the problem is that each example can be limited by the specifics of the shaders it needs. So one can enable experimental translation on Metal, for example, another can not. So for now I think having them separate is better.

grovesNL replied:
Yeah for now it's probably fine like this – it might be nice to have a helper once it's ready on all examples

grovesNL commented on examples/cube/main.rs:
```diff
@@ -274,19 +271,27 @@ impl framework::Example for Example {
+        }
+        let shader = device.create_shader_module(&wgpu::ShaderModuleDescriptor {
+            label: None,
+            source: wgpu::ShaderSource::Wgsl(Cow::Borrowed(include_str!("shader.wgsl"))),
```
nit: this will break the examples on the web backend for a while, though maybe we could start generating SPIR-V internally soon or temporarily keep the spv files as a fallback path for the web

kvark replied:
Good point! I'll hook Naga up to the web side and convert it.

grovesNL reviewed (approved):
Awesome :+1: great to see this coming along

kvark commented:
Going to land this right after #734

kvark commented:
A little functional changes here
  - switched "boids" to use read-only storage for the source array
  - switched "shadow" to use the storage with runtime-sized array instead of a uniform structure

bors r=grovesNL

dneto0 commented on examples/shadow/shader.wgsl:
```diff
@@ -0,0 +1,114 @@
+
+// fragment shader
+
+[[block]]
```
FYI. It's invalid to declare an array of block-decorated struct.
WGSL doesn't have array-of-buffers, but Vulkan SPIR-V does

kvark replied:
Right. At the time this was written, our `[[block]]` semantics was a bit different. Today, there is no `[[block]]` on it https://github.com/gfx-rs/wgpu-rs/blob/69be1ff88a073b3a7f9470044834e75c22f7b4ad/examples/shadow/shader.wgsl#L54
